### ISSUE TITLE
[CT-2491] [Bug] Default `--target-path` should be relative to `--project-dir`, rather than current working directory

### ISSUE THREAD

**Issue by OwenKephart** [dbt-labs/dbt-core#7465]
### Is this a new bug in dbt-core?

- [X] I believe this is a new bug in dbt-core
- [X] I have searched the existing issues, and I could not find an existing issue for this bug

### Current Behavior

Executing a command that targets a project in a specified `--project-dir` results in the json artifacts being written to `current-working-directory/target/` rather than `project-dir/target/`.

### Expected Behavior

In previous versions, specifying a `target-path` in your dbt_project.yml was relative to the project directory, rather than relative to the current working directory from which the command was executed.

### Steps To Reproduce

In a new virtual environment (just using duckdb so I don't have to get docker involved):

```
pip install dbt-core==1.5.0rc2 dbt-duckdb==1.5.0rc1

gh repo clone dbt-labs/jaffle_shop

cd jaffle_shop

echo "jaffle_shop:
  outputs:
   dev:
     type: duckdb
     path: /tmp/dbt.duckdb
  target: dev" > profiles.yml
  
 cd some/other/path
 
 dbt build --profiles-dir ~/jaffle_shop --project-dir ~/jaffle_shop
 
 ls some/other/path 
```

At this point, you will observe that a new `target` directory has been created in `some/other/path`, and no target directory exists in `~/jaffle_shop`.

### Environment

```markdown
- OS: Ventura 13.12
- Python: 3.9.10
- dbt: 1.5.0rc2
```

### Which database adapter are you using with dbt?

other (mention it in "Additional Context")

### Additional Context

For this example, I'm using the duckdb adapter just for ease of reproducibility, but I don't think that has an impact on this specific behavior.

**Comment by OwenKephart** [comment]
@jtcohen6 Thanks for the quick response! (and oops on missing that bit in the migration guide).

The reasoning for removing `target-path` laid out in that PR definitely makes sense (and it's great that this can now be customized per-invocation! super useful).

Setting the `DBT_TARGET_PATH` env var has the nice property of being valid to set regardless of the user's dbt-core version (working on the `dagster-dbt` integration), so I'll swap over to that. Previously, I'd just been setting the subprocess's working directory to the project dir 😬 

Of course, that will all become unnecessary once we port over to the `dbtRunner` stuff, which I'm very excited to do :)

I do still feel like the default behavior here should be to write target files to `dbt_project_dir/target/` rather than `current_working_directory/target/`, but we're unblocked on this for now.

**Comment by jtcohen6** [comment]
> Of course, that will all become unnecessary once we port over to the `dbtRunner` stuff, which I'm very excited to do :)

Right on - if you didn't catch this in the docs, it would also be possible to set `target_path` by passing it as a kwarg to the `invoke` method: https://docs.getdbt.com/reference/programmatic-invocations#overriding-parameters

> I do still feel like the default behavior here should be to write target files to `dbt_project_dir/target/` rather than `current_working_directory/target/`, but we're unblocked on this for now.

This makes sense to me, and it looks like I said the same over in https://github.com/dbt-labs/dbt-core/issues/6882#issuecomment-1426089712. I'll keep this issue open accordingly.

**Comment by iknox-fa** [comment]
Additional context from an internal Slack convo:
https://dbt-labs.slack.com/archives/C0131TY7EEA/p1682626884003739

TL;DR--
When `target_path` is not explicitly set via an env_var AND dbt is invoked from a working directory that isn't the project's root directory some artifacts are written to `project_root/target` and some are written to `cwd/target`

**Comment by jtcohen6** [comment]
Upgrading this to a `regression` given that it's causing surprising/difficult behavior for several users/plugins

**Comment by dwreeves** [comment]
> Additional context from an internal Slack convo: https://dbt-labs.slack.com/archives/C0131TY7EEA/p1682626884003739
> 
> TL;DR-- When `target_path` is not explicitly set via an env_var AND dbt is invoked from a working directory that isn't the project's root directory some artifacts are written to `project_root/target` and some are written to `cwd/target`

I am running into this same issue but with the `--target-path` in the CLI. We are passing in `./dags/dbt/target` as our target path. It appears to be the `partial_parse.msgpack` cache thingy is written to `project_root/target` and everything else is set to `cwd/target`. Note this file is written after everything runs, so the issue is a change of directory occurs at some point.

Note that I was able to solve the issue by replacing `--target-path ./dags/dbt/target` with `--target-path $(realpath ./dags/dbt/target)`.

**Comment by dwreeves** [comment]
At least as far as the CLI is concerned, would it be possible to just fix the issue using a callback in the `click.Option` that wraps the user input with `os.path.abspath()`? If this is an acceptable solution, I can open a PR that does this.

**Comment by jaklan** [comment]
We have just encountered the same issue after the upgrade to `1.5` which broke our CI pipelines.

 - when having legacy `target-path: "target"` in `common/dbt/dbt_project.yml`, the below command:
    ```bash
    dbt compile --project-dir=common/dbt --profiles-dir=common/dbt
    ```
    results in:
    - `partial_parse.msgpack` written to `common/dbt/target`
    - `compiled`, `graph.gpickle`, and `manifest.json` written to `target`

- when I remove the `target-path` config - the result is the same
- when I both remove `target-path` and add `--target-path=target`- the result is the same
- the only working solution is `--target-path=<ABSOLUTE_PATH>/target`

**Comment by jaklan** [comment]
We have just discovered the same issue happens when using `--defer` flag: `--defer artifacts` doesn't look at `<project-dir>/artifacts` anymore, but in `<cwd>/artifacts`. That's ofc unexpected and the previous behaviour should be preserved.

**Comment by jtcohen6** [comment]
The fix for this will be in v1.5.2 (next patch)

**Comment by chamini2** [comment]
Sounds great! Thank you

**Comment by britt-allen** [comment]
When will v1.5.2 be released?
